lib/maths/unittest/CBoostedTreeTest.cc

```diff
@@ -1647,21 +1647,96 @@ BOOST_AUTO_TEST_CASE(testPersistRestore) {
-                           .restoreFor(*frame, cols - 1);
+    // Restore.
+    auto boostedTree =
+        maths::CBoostedTreeFactory::constructFromString(persistOnceState).restoreFor(*frame, cols - 1);
     {
-        core::CJsonStatePersistInserter inserter(persistTwiceSStream);
+        // We need to call the inserter destructor to finish writing the state.
+        core::CJsonStatePersistInserter inserter(persistTwiceState);
         boostedTree->acceptPersistInserter(inserter);
-        persistTwiceSStream.flush();
+        persistTwiceState.flush();
+    }
+    LOG_DEBUG(<< "State " << persistOnceState.str());
+    LOG_DEBUG(<< "State after restore " << persistTwiceState.str());
+    BOOST_REQUIRE_EQUAL(persistOnceState.str(), persistTwiceState.str());
+}
+
+BOOST_AUTO_TEST_CASE(testPersistRestoreDuringInitialization) {
+
+    // Grab checkpoints during initialization and check they all produce the
+    // same result after restoring.
+
+    using TSStreamVec = std::vector<std::stringstream>;
+
+    std::size_t rows{50};
+    std::size_t cols{3};
+    std::size_t capacity{50};
+    test::CRandomNumbers rng;
+
+    // Generate completely random data.
+    TDoubleVecVec x(cols);
+    for (std::size_t i = 0; i < cols; ++i) {
+        rng.generateUniformSamples(0.0, 10.0, rows, x[i]);
+    }
+
+    auto makeFrame = [&] {
+        auto frame = core::makeMainStorageDataFrame(cols, capacity).first;
+        for (std::size_t i = 0; i < rows; ++i) {
+            frame->writeRow([&](core::CDataFrame::TFloatVecItr column, std::int32_t&) {
+                for (std::size_t j = 0; j < cols; ++j, ++column) {
+                    *column = x[j][i];
+                }
+            });
+        }
+        frame->finishWritingRows();
+        return frame;
+    };
+
+    TSStreamVec checkpoints;
+
+    auto writeCheckpoint = [&checkpoints](maths::CBoostedTree::TPersistFunc persist) {
+        std::stringstream state;
+        {
+            // We need to call the inserter destructor to finish writing the state.
+            core::CJsonStatePersistInserter inserter(state);
+            persist(inserter);
+            state.flush();
+        }
+        checkpoints.push_back(std::move(state));
+    };
+
+    std::ostringstream expectedState;
+    {
+        auto frame = makeFrame();
+        auto boostedTree = maths::CBoostedTreeFactory::constructFromParameters(
+                               1, std::make_unique<maths::boosted_tree::CMse>())
+                               .numberFolds(2)
+                               .maximumNumberTrees(2)
+                               .maximumOptimisationRoundsPerHyperparameter(3)
+                               .trainingStateCallback(writeCheckpoint)
+                               .buildFor(*frame, cols - 1);
+        core::CJsonStatePersistInserter inserter(expectedState);
+        boostedTree->acceptPersistInserter(inserter);
+        expectedState.flush();
+    }
+
```

Comment:
Nice test overall, I think you need to check  that `checkpoints` contains the expected number of elements, otherwise mot may be empty and the loop below is useless.

Reply:
Great point!

Reply:
See [ea2299c](https://github.com/elastic/ml-cpp/pull/1300/commits/ea2299cdcea8c00055b68725b31dc153580fafc2).

Reply:
Note, I decided to just add a test that there is at least one checkpoint so we don't fail on changes to when we checkpoint.